File: espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/w3c/KeyInputStateTest.java
package io.appium.espressoserver.test.helpers.w3c;


import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;

import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;


public class KeyInputStateTest {

    private KeyInputState keyInputStateOne;
    private KeyInputState keyInputStateTwo;

    @Before
    public void before() {
        keyInputStateOne = new KeyInputState();
        keyInputStateTwo = new KeyInputState();
    }

    @Test
    public void shouldSetAltToTrueIfJustOneIsTrue() {
        keyInputStateOne.setAlt(true);
        keyInputStateTwo.setAlt(false);
        KeyInputState[] keyInputStates = new KeyInputState[]{keyInputStateOne, keyInputStateTwo};
        KeyInputState aggregateKeyInputState = KeyInputState.getGlobalKeyState(Arrays.asList(keyInputStates));
        assertTrue(aggregateKeyInputState.isAlt());
    }

    @Test
    public void shouldSetAltToFalseIfAllAreFalse() {
        keyInputStateOne.setAlt(false);
        keyInputStateTwo.setAlt(false);
        KeyInputState[] keyInputStates = new KeyInputState[]{keyInputStateOne, keyInputStateTwo};
        KeyInputState aggregateKeyInputState = KeyInputState.getGlobalKeyState(Arrays.asList(keyInputStates));
        assertFalse(aggregateKeyInputState.isAlt());
    }

    @Test
    public void shouldTestSeveralKeyStates() {
        keyInputStateOne.setAlt(true);
        keyInputStateOne.setCtrl(true);
        keyInputStateOne.setShift(false);
        keyInputStateOne.setMeta(false);

        keyInputStateTwo.setAlt(false);
        keyInputStateTwo.setCtrl(true);
        keyInputStateTwo.setShift(false);
        keyInputStateTwo.setMeta(true);

        KeyInputState[] keyInputStates = new KeyInputState[]{keyInputStateOne, keyInputStateTwo};
        KeyInputState aggregateKeyInputState = KeyInputState.getGlobalKeyState(Arrays.asList(keyInputStates));
        assertTrue(aggregateKeyInputState.isAlt());
        assertTrue(aggregateKeyInputState.isCtrl());
        assertFalse(aggregateKeyInputState.isShift());
        assertTrue(aggregateKeyInputState.isMeta());
    }

    @Test
    public void shouldCombinePressed() {
        keyInputStateOne.addPressed("a");
        keyInputStateOne.addPressed("b");
        keyInputStateOne.addPressed("c");
        keyInputStateOne.addPressed("d");
        keyInputStateTwo.addPressed("d");
        keyInputStateTwo.addPressed("e");
        keyInputStateTwo.addPressed("f");

        KeyInputState[] keyInputStates = new KeyInputState[]{keyInputStateOne, keyInputStateTwo};
        KeyInputState aggregateKeyInputState = KeyInputState.getGlobalKeyState(Arrays.asList(keyInputStates));
        assertTrue(aggregateKeyInputState.isPressed("a"));
        assertTrue(aggregateKeyInputState.isPressed("b"));
        assertTrue(aggregateKeyInputState.isPressed("c"));
        assertTrue(aggregateKeyInputState.isPressed("d"));
        assertTrue(aggregateKeyInputState.isPressed("e"));
        assertTrue(aggregateKeyInputState.isPressed("f"));
        assertTrue(!aggregateKeyInputState.isPressed("g"));
    }
}

File: espresso-server/app/src/test/java/io/appium/espressoserver/test/helpers/w3c/KeyInputStateTest.kt
package io.appium.espressoserver.test.helpers.w3c

import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState
import io.appium.espressoserver.lib.helpers.w3c.state.getGlobalKeyState
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class KeyInputStateTest {
    private var keyInputStateOne: KeyInputState? = null
    private var keyInputStateTwo: KeyInputState? = null

    @Before
    fun before() {
        keyInputStateOne = KeyInputState()
        keyInputStateTwo = KeyInputState()
    }

    @Test
    fun shouldSetAltToTrueIfJustOneIsTrue() {
        keyInputStateOne!!.isAlt = true
        keyInputStateTwo!!.isAlt = false
        val keyInputStates = arrayOf(keyInputStateOne, keyInputStateTwo)
        val aggregateKeyInputState = getGlobalKeyState(listOfNotNull(*keyInputStates))
        Assert.assertTrue(aggregateKeyInputState.isAlt)
    }

    @Test
    fun shouldSetAltToFalseIfAllAreFalse() {
        keyInputStateOne!!.isAlt = false
        keyInputStateTwo!!.isAlt = false
        val keyInputStates = arrayOf(keyInputStateOne, keyInputStateTwo)
        val aggregateKeyInputState = getGlobalKeyState(listOfNotNull(*keyInputStates))
        Assert.assertFalse(aggregateKeyInputState.isAlt)
    }

    @Test
    fun shouldTestSeveralKeyStates() {
        keyInputStateOne!!.isAlt = true
        keyInputStateOne!!.isCtrl = true
        keyInputStateOne!!.isShift = false
        keyInputStateOne!!.isMeta = false
        keyInputStateTwo!!.isAlt = false
        keyInputStateTwo!!.isCtrl = true
        keyInputStateTwo!!.isShift = false
        keyInputStateTwo!!.isMeta = true
        val keyInputStates = arrayOf(keyInputStateOne, keyInputStateTwo)
        val aggregateKeyInputState = getGlobalKeyState(listOfNotNull(*keyInputStates))
        Assert.assertTrue(aggregateKeyInputState.isAlt)
        Assert.assertTrue(aggregateKeyInputState.isCtrl)
        Assert.assertFalse(aggregateKeyInputState.isShift)
        Assert.assertTrue(aggregateKeyInputState.isMeta)
    }

    @Test
    fun shouldCombinePressed() {
        keyInputStateOne!!.addPressed("a")
        keyInputStateOne!!.addPressed("b")
        keyInputStateOne!!.addPressed("c")
        keyInputStateOne!!.addPressed("d")
        keyInputStateTwo!!.addPressed("d")
        keyInputStateTwo!!.addPressed("e")
        keyInputStateTwo!!.addPressed("f")
        val keyInputStates = arrayOf(keyInputStateOne, keyInputStateTwo)
        val aggregateKeyInputState = getGlobalKeyState(listOfNotNull(*keyInputStates))
        Assert.assertTrue(aggregateKeyInputState.isPressed("a"))
        Assert.assertTrue(aggregateKeyInputState.isPressed("b"))
        Assert.assertTrue(aggregateKeyInputState.isPressed("c"))
        Assert.assertTrue(aggregateKeyInputState.isPressed("d"))
        Assert.assertTrue(aggregateKeyInputState.isPressed("e"))
        Assert.assertTrue(aggregateKeyInputState.isPressed("f"))
        Assert.assertFalse(aggregateKeyInputState.isPressed("g"))
    }
}